File: java/structures/PersistentTree.java
package structures;

// https://en.wikipedia.org/wiki/Persistent_data_structure
public class PersistentTree {

    public static class Node {
        Node left, right;
        int sum;

        Node(int value) {
            sum = value;
        }

        Node(Node left, Node right) {
            this.left = left;
            this.right = right;
            if (left != null)
                sum += left.sum;
            if (right != null)
                sum += right.sum;
        }
    }

    public static Node build(int left, int right) {
        if (left == right)
            return new Node(0);
        int mid = (left + right) >> 1;
        return new Node(build(left, mid), build(mid + 1, right));
    }

    public static int sum(int from, int to, Node root, int left, int right) {
        if (from > right || left > to)
            return 0;
        if (from <= left && right <= to)
            return root.sum;
        int mid = (left + right) >> 1;
        return sum(from, to, root.left, left, mid) + sum(from, to, root.right, mid + 1, right);
    }

    public static Node set(int pos, int value, Node root, int left, int right) {
        if (left == right)
            return new Node(value);
        int mid = (left + right) >> 1;
        return pos <= mid ?
                new Node(set(pos, value, root.left, left, mid), root.right) :
                new Node(root.left, set(pos, value, root.right, mid + 1, right));
    }

    // Usage example
    public static void main(String[] args) {
        int n = 10;
        Node t1 = build(0, n - 1);
        Node t2 = set(0, 1, t1, 0, n - 1);
        System.out.println(0 == sum(0, 9, t1, 0, n - 1));
        System.out.println(1 == sum(0, 9, t2, 0, n - 1));
    }
}

File: rust/structures/persistent_tree.rs
use std::rc::Rc;

pub struct Node {
    left: Option<Rc<Node>>,
    right: Option<Rc<Node>>,
    sum: i32,
}

impl Node {
    fn new(sum: i32) -> Self {
        Self {
            left: None,
            right: None,
            sum,
        }
    }

    fn new_parent(left: &Rc<Node>, right: &Rc<Node>) -> Self {
        Self {
            left: Some(left.clone()),
            right: Some(right.clone()),
            sum: left.sum + right.sum,
        }
    }
}

pub struct PersistentTree;
impl PersistentTree {
    pub fn build(left: i32, right: i32) -> Rc<Node> {
        let node = if left == right {
            Node::new(0)
        } else {
            let mid = (left + right) >> 1;
            Node::new_parent(&Self::build(left, mid), &Self::build(mid + 1, right))
        };
        Rc::new(node)
    }

    pub fn sum(from: i32, to: i32, root: &Node, left: i32, right: i32) -> i32 {
        if from > right || left > to {
            0
        } else if from <= left && right <= to {
            root.sum
        } else {
            let mid = (left + right) >> 1;
            Self::sum(from, to, root.left.as_ref().unwrap(), left, mid)
                + Self::sum(from, to, root.right.as_ref().unwrap(), mid + 1, right)
        }
    }

    pub fn set(pos: i32, value: i32, root: &Node, left: i32, right: i32) -> Rc<Node> {
        let node = if left == right {
            Node::new(value)
        } else {
            let mid = (left + right) >> 1;
            if pos <= mid {
                Node::new_parent(
                    &Self::set(pos, value, root.left.as_ref().unwrap(), left, mid),
                    root.right.as_ref().unwrap(),
                )
            } else {
                Node::new_parent(
                    root.left.as_ref().unwrap(),
                    &Self::set(pos, value, root.right.as_ref().unwrap(), mid + 1, right),
                )
            }
        };
        Rc::new(node)
    }
}

#[cfg(test)]
mod tests {
    use crate::structures::persistent_tree::PersistentTree;

    #[test]
    fn basic_test() {
        let n = 10;
        let t1 = PersistentTree::build(0, n - 1);
        let t2 = PersistentTree::set(0, 1, &t1, 0, n - 1);
        assert_eq!(PersistentTree::sum(0, n - 1, &t1, 0, n - 1), 0);
        assert_eq!(PersistentTree::sum(0, n - 1, &t2, 0, n - 1), 1);
    }
}
